### PR TITLE
[mypyc] Add a str.format specializer which only supports empty brackets

### PR DESCRIPTION
**Pull request by 97littleleaf11**
### Description

This PR adds a str.format specializer which only supports empty brackets. Also, the detection code can deal with the bracket literals, for example, "{{}}".

The specializer first split the formatting string by "{}". Then it replaces the brackets by `PyObject_Str`. Finally, all these separated substrings would be sent to a C helper function to generate a joint one.

## Test Plan

* An IRbuild test
* A run test

mypyc_benchmark:
```
running str_format_format_method
..........
interpreted: 0.597254s (avg of 5 iterations; stdev 1.7%)
compiled:    0.312174s (avg of 5 iterations; stdev 2.9%)

compiled is 1.913x faster
```

### CODE REVIEW

File: mypyc/irbuild/specialize.py [hunk 1]
```diff
@@ -356,3 +357,45 @@ def translate_dict_setdefault(
...
+        builder: IRBuilder, expr: CallExpr, callee: RefExpr) -> Optional[Value]:
+    if isinstance(callee, MemberExpr) and isinstance(callee.expr, StrExpr):
+        # TODO
+        # Only consider simplest situation here
+        format_str = callee.expr.value
+        if not str_format_check_helper(format_str):
+            return None
+
+        literals = [builder.load_str(x) for x in format_str.split("{}")]
+        variables = [builder.call_c(str_op, [builder.accept(x)], expr.line) for x in expr.args]
```
Comment (JukkaL) [inline mypyc/irbuild/specialize.py:373]: You can skip the `str_op` call if the type of the argument is `str_rprimitive`.

File: mypyc/irbuild/specialize.py [hunk 2]
```diff
@@ -356,3 +357,45 @@ def translate_dict_setdefault(
                               [callee_dict, key_val, data_type],
                               expr.line)
     return None
+
+
+@specialize_function('format', str_rprimitive)
+def translate_str_format(
+        builder: IRBuilder, expr: CallExpr, callee: RefExpr) -> Optional[Value]:
+    if isinstance(callee, MemberExpr) and isinstance(callee.expr, StrExpr):
```
Comment (JukkaL) [inline mypyc/irbuild/specialize.py:365]: Check that all arguments are positional, as this only works for positional args.

File: mypyc/irbuild/specialize.py [hunk 3]
```diff
@@ -356,3 +357,45 @@ def translate_dict_setdefault(
...
+        result_list[0] = Integer(result_len, c_int_rprimitive, expr.line)
+        result_list[1::2] = literals
+        result_list[2::2] = variables
+        return builder.call_c(str_build_op, result_list, expr.line)
+    return None
+
+
+# A helper function for checking whether the format_str can be optimized by
+# translate_str_format().
+def str_format_check_helper(format_str: str) -> bool:
```
Comment (JukkaL) [inline mypyc/irbuild/specialize.py:390]: Style nit: "helper" is not a very informative name. Try to name functions based on what they do. For example, here a better name could be `can_optimize_format`.

File: mypyc/irbuild/specialize.py [hunk 4]
```diff
@@ -356,3 +357,45 @@ def translate_dict_setdefault(
...
+        # The first parameter is the total size of the following PyObject* merged from
+        # two lists alternatively.
+        result_list[0] = Integer(result_len, c_int_rprimitive, expr.line)
+        result_list[1::2] = literals
+        result_list[2::2] = variables
+        return builder.call_c(str_build_op, result_list, expr.line)
+    return None
+
+
+# A helper function for checking whether the format_str can be optimized by
```
Comment (JukkaL) [inline mypyc/irbuild/specialize.py:388]: Style nit: Use docstring instead of a comment before a function for consistency. If you can find a more informative name for the function, the comment may become redundant.

File: mypyc/irbuild/specialize.py [hunk 5]
```diff
@@ -356,3 +357,45 @@ def translate_dict_setdefault(
...
+
+        # Allocate space for the parameter list of CPyStr_Build().
+        result_len = len(literals) + len(variables)
+        result_list: List[Value] = [Integer(0, c_int_rprimitive)] * (result_len + 1)
+
+        # The first parameter is the total size of the following PyObject* merged from
+        # two lists alternatively.
+        result_list[0] = Integer(result_len, c_int_rprimitive, expr.line)
+        result_list[1::2] = literals
+        result_list[2::2] = variables
```
Comment (JukkaL) [inline mypyc/irbuild/specialize.py:383]: These slice assignments are somewhat difficult to reason about (a bit too clever, I think). I'd prefer a simple for loop, possible with `zip()`. This way you should be able to build the list incrementally using `append` instead of first creating it and then mutating it. Appending to a list is more Pythonic than creating a list first with dummy values and then assigning to it.

File: mypyc/irbuild/specialize.py [hunk 6]
```diff
@@ -356,3 +357,45 @@ def translate_dict_setdefault(
...
+        variables = [builder.call_c(str_op, [builder.accept(x)], expr.line) for x in expr.args]
+
+        # Allocate space for the parameter list of CPyStr_Build().
+        result_len = len(literals) + len(variables)
+        result_list: List[Value] = [Integer(0, c_int_rprimitive)] * (result_len + 1)
+
+        # The first parameter is the total size of the following PyObject* merged from
+        # two lists alternatively.
+        result_list[0] = Integer(result_len, c_int_rprimitive, expr.line)
+        result_list[1::2] = literals
```
Comment (JukkaL) [inline mypyc/irbuild/specialize.py:382]: You can skip empty literals, as they will have no effect and just slow things down a bit at runtime.

**Review by JukkaL — COMMENTED** [review]
I'm glad to see the performance improvement! Left some mostly minor comments. Once you handle `{{` and `}}` this seems enough features for a first iteration.

**Comment by 97littleleaf11** [comment]
Add several optimization:

- Skip empty literals
- Skip str_op on str_rprimitive

running str_format_format_method
..........
interpreted: 0.576623s (avg of 5 iterations; stdev 0.47%)
compiled:    0.280255s (avg of 5 iterations; stdev 0.63%)

compiled is 2.057x faster

**Comment by JukkaL** [comment]
The build is failing. Can you have a look?

**Comment by 97littleleaf11** [comment]
@JukkaL I fixed it by replacing "{{" with "{"

File: mypyc/test-data/run-strings.test [hunk 7]
```diff
@@ -178,3 +178,22 @@ def test_str_to_bool() -> None:
...
+    assert "{}".format(name) == "Eric"
+    assert "{}! Hi!".format(name) == "Eric! Hi!"
+    assert "{}, Hi, {}".format(name, name) == "Eric, Hi, Eric"
+    assert "Hi! {}".format(name) == "Hi! Eric"
+    assert "Hi, I'm {}. I'm {}.".format(name, age) == "Hi, I'm Eric. I'm 14."
+
+    assert "{{}}{}".format(name) == "{}Eric"
+    assert "Hi! {{{}}}".format(name) == "Hi! {Eric}"
+    assert "Hi! {{ {}".format(name) == "Hi! { Eric"
+    assert "Hi! {{ {} }}}}".format(name) == "Hi! { Eric }}"
```
Comment (JukkaL) [inline mypyc/test-data/run-strings.test:198]: Please add more tests. We want fairly complete test coverage to make sure nothing has regressed. It's important both valid and error cases produce the same output when optimized. For example, any exceptions should be identical.

Examples (but there are probably more):
* Simplest possible case: `''.format()`
* Plain text only: `'xyz'.format()`
* Extra positional arguments: `'a{}b'.format(1, 2, 3)`
* Too few arguments
* Various format specifiers that are supported within `{...}` (these should still work, even if they aren't optimized)
* Keyword arguments to `format`
* `*args` to `format`
* Invalid format string

File: mypyc/test-data/irbuild-str.test [hunk 8]
```diff
@@ -159,3 +159,20 @@ L2:
 L3:
     unreachable
 
+[case testStrFormatMethod]
```
Comment (JukkaL) [inline mypyc/test-data/irbuild-str.test:162]: Add irbuild tests for skipping `PyObject_Str` for `str` values, skipping empty strings and `{{` / `}}`.

File: mypyc/irbuild/specialize.py [hunk 9]
```diff
@@ -356,3 +358,53 @@ def translate_dict_setdefault(
...
+        builder: IRBuilder, expr: CallExpr, callee: RefExpr) -> Optional[Value]:
+    if (isinstance(callee, MemberExpr) and isinstance(callee.expr, StrExpr)
+            and expr.arg_kinds.count(ARG_POS) == len(expr.arg_kinds)):
+
+        format_str = callee.expr.value
+        if not can_optimize_format(format_str):
+            return None
+
+        format_str = format_str.replace("{{", "{")
+        format_str = format_str.replace("}}", "}")
```
Comment (JukkaL) [inline mypyc/irbuild/specialize.py:374]: This isn't right? `{{}}` gets translated to `{}`, and we'll split it below, even though we shouldn't do it?

File: mypyc/irbuild/specialize.py [hunk 10]
```diff
@@ -356,3 +358,53 @@ def translate_dict_setdefault(
...
+        # The str.split() always generates one more element
+        if literals[-1]:
+            result_list.append(builder.load_str(literals[-1]))
+
+        result_list[0] = Integer(len(result_list) - 1, c_int_rprimitive)
+        return builder.call_c(str_build_op, result_list, expr.line)
+    return None
+
+
+def can_optimize_format(format_str: str) -> bool:
```
Comment (JukkaL) [inline mypyc/irbuild/specialize.py:397]: Please add unit tests for this, since it has pretty tricky logic. Also consider various error cases and edge cases.

**Review by JukkaL — COMMENTED** [review]
Thanks for the updates! Left various suggestions about more test coverage. Right now test coverage is still a bit limited, and there may be some bugs still lurking in the code.

File: mypyc/irbuild/specialize.py (hunk 9 above)
Comment (97littleleaf11) [reply mypyc/irbuild/specialize.py:374]: Since we would support many operations other than empty brackets, I'll add a helper function for parsing instead of using `replace` and `split`.

File: mypyc/test-data/run-strings.test (hunk 7 above)
Comment (97littleleaf11) [reply mypyc/test-data/run-strings.test:198]: Extra positional arguments case this compile error: `error: Not all arguments converted during string formatting`

Comment (JukkaL) [reply mypyc/test-data/run-strings.test:303]: These test cases could still be useful. Can you double check if they aren't covered:
* `'{}{}'.format(1, 2)` (empty literals)
* `'x{}y{}'.format(*args)` (`*args`)

**Review by JukkaL — COMMENTED** [review]
Thanks for the updates! This is almost ready to merge. Just a few remaining comments.

**Review by JukkaL — APPROVED** [review]
LGTM!
